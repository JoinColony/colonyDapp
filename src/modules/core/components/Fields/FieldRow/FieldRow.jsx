/* @flow */

import React, { Component } from 'react';
import type { ComponentType } from 'react';
import type { FieldProps } from 'redux-form';
import type { Appearance } from '$types/css';

import type { FieldComponentProps, Option, Utils } from '../flowTypes';

type Props = FieldProps & {
  appearance?: Appearance,
  elementOnly?: boolean,
  fieldComponent: ComponentType<FieldComponentProps<>>,
  help: string,
  id: string,
  immediateValidation?: boolean,
  label: string,
  name: string,
  options?: Array<Option>,
  placeholder: string,
  utils: Utils,
  title: string,
};

class FieldRow extends Component<Props> {
  reset: () => void;
  static displayName = 'core.Field.FieldRow';
  constructor(props: Props) {
    super(props);
    this.reset = this.reset.bind(this);
  }
  reset() {
    const { input: { onChange } } = this.props;
    onChange('');
  }
  render() {
    const {
      appearance,
      custom,
      elementOnly,
      fieldComponent: FieldComponent,
      help,
      id,
      immediateValidation,
      input,
      label,
      meta,
      options,
      placeholder,
      utils,
      title,
      ...props
    } = this.props;

    const { touched, pristine, error } = meta;
    const hasError = !!(error && (touched || (immediateValidation && !pristine)));

    const inputProps = {
      id,
      placeholder,
      title: hasError ? (error || title) : title,
    };

    return (
      <FieldComponent
        appearance={appearance}
        custom={custom}
        elementOnly={elementOnly}
        error={error}
        hasError={hasError}
        help={help}
        input={input}
        inputProps={inputProps}
        label={label}
        meta={meta}
        options={options}
        passthroughProps={props}
        utils={{
          ...utils,
          reset: utils.reset || this.reset,
        }}
      />
    );
  }
}
export default FieldRow;
