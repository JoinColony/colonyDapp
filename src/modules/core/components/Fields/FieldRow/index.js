/* @flow */

export { default } from './FieldRow.jsx';
