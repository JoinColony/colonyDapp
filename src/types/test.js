/* @flow */

import {
  Address,
  ENSName,
  type AddressType,
  type ENSNameType,
} from './strings';

const takesENSName = (arg: ENSNameType) => arg;
const takesAddress = (arg: AddressType) => arg;

const name = ENSName('ens name');
const address = Address('0x123');

takesENSName(name);
takesAddress(address);

/*
 * The following should produce flow errors:
 */
// Address('not a hex string');
// takesAddress('0x123'); // just a string
// takesENSName('just a string');
