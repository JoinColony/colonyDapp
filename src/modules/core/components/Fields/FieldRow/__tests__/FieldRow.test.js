/* eslint-env jest */

import React from 'react';
import toJson from 'enzyme-to-json';

import { shallowWithIntl } from '$utils/test';

import FieldRow from '../FieldRow.jsx';

const TestFieldComponent = () => <div />;

describe('FieldRow component', () => {
  test('Renders initial component', () => {
    const wrapper = shallowWithIntl((
      <FieldRow
        name="test"
        fieldComponent={TestFieldComponent}
        meta={{ touched: false, pristine: false, error: false, active: false }}
        utils={{}}
        input={{ checked: false, disabled: false, onChange: jest.fn() }}
      />
    ));
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('With hasError, error shows instead of title', () => {
    const wrapper = shallowWithIntl((
      <FieldRow
        title="AwesomeComponent"
        name="test"
        fieldComponent={TestFieldComponent}
        meta={{ touched: true, pristine: false, error: 'You made a booboo', active: false }}
        utils={{}}
      />
    ));
    expect(wrapper.prop('inputProps').title).toBe('You made a booboo');
  });
});
