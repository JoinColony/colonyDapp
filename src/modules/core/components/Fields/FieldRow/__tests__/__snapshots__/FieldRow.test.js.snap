// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldRow component Renders initial component 1`] = `
<TestFieldComponent
  error={false}
  hasError={false}
  input={
    Object {
      "checked": false,
      "disabled": false,
      "onChange": [Function],
    }
  }
  inputProps={
    Object {
      "id": undefined,
      "placeholder": undefined,
      "title": undefined,
    }
  }
  meta={
    Object {
      "active": false,
      "error": false,
      "pristine": false,
      "touched": false,
    }
  }
  passthroughProps={
    Object {
      "intl": Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatHTMLMessage": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatPlural": [Function],
        "formatRelative": [Function],
        "formatTime": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralFormat": [Function],
          "getRelativeFormat": [Function],
        },
        "locale": "en",
        "messages": Object {
          "button.add": "Add",
          "button.back": "Back",
          "button.cancel": "Cancel",
          "button.confirm": "Confirm",
          "button.dismiss": "Dismiss",
          "button.next": "Next",
          "button.ok": "Okay",
          "button.remove": "Remove",
          "home": "Home",
          "input.resetText": "Cancel",
          "label.description": "Description",
          "label.email": "Email address",
          "label.name": "Name",
          "label.optional": "Optional",
          "label.password": "Password",
          "placeholder.email": "handle@domain.com",
          "placeholder.password": "Password",
          "role.admin": "Admin",
          "role.disabled": "Disabled",
          "role.member": "Member",
          "role.owner": "Owner",
        },
        "now": [Function],
        "textComponent": "span",
      },
      "name": "test",
    }
  }
  utils={
    Object {
      "reset": [Function],
    }
  }
/>
`;
